File: win/src/UpdateChecker.cs
using System;
using System.Globalization;
using System.IO;
using System.Net;
using System.Reflection;
using System.Threading;
using System.Windows.Forms;
using System.Web.Script.Serialization;

namespace Sanity
{
    public static class UpdateChecker
    {
        private static readonly TimeSpan CheckInterval = TimeSpan.FromDays(7);
        private static int _running;

        public static void RunAsync(AppConfig config, SynchronizationContext ui)
        {
            if (config == null || ui == null)
                return;
            if (!ShouldStart(config))
                return;

            ThreadPool.QueueUserWorkItem(_ =>
            {
                RegexRulesCatalog remoteRules = null;
                DateTime? remoteReleaseDate = null;

                try
                {
                    remoteRules = DefaultRules.FetchRemote();
                }
                catch
                {
                }

                try
                {
                    remoteReleaseDate = FetchReleaseFileDate();
                }
                catch
                {
                }

                ui.Post(__ =>
                {
                    try
                    {
                        Prompt(config, remoteRules, remoteReleaseDate);
                    }
                    finally
                    {
                        Interlocked.Exchange(ref _running, 0);
                    }
                }, null);
            });
        }

        /// <summary>
        /// Blocks while fetching and prompting — for short-lived --open processes.
        /// </summary>
        public static void RunSync(AppConfig config)
        {
            if (config == null || !ShouldStart(config))
                return;

            try
            {
                RegexRulesCatalog remoteRules = null;
                DateTime? remoteReleaseDate = null;

                try
                {
                    remoteRules = DefaultRules.FetchRemote();
                }
                catch
                {
                }

                try
                {
                    remoteReleaseDate = FetchReleaseFileDate();
                }
                catch
                {
                }

                Prompt(config, remoteRules, remoteReleaseDate);
            }
            finally
            {
                Interlocked.Exchange(ref _running, 0);
            }
        }

        private static bool ShouldStart(AppConfig config)
        {
            if (!config.UpdatesEnabled)
                return false;
            if (!IsDue(config))
                return false;
            return Interlocked.CompareExchange(ref _running, 1, 0) == 0;
        }

        private static bool IsDue(AppConfig config)
        {
            if (!config.LastUpdateCheck.HasValue)
                return true;

            return DateTime.UtcNow - config.LastUpdateCheck.Value.ToUniversalTime() >= CheckInterval;
        }

        private static void Prompt(
            AppConfig config,
            RegexRulesCatalog remoteRules,
            DateTime? remoteReleaseDate)
        {
            try
            {
                if (remoteRules != null && remoteRules.Version > config.RulesVersion)
                {
                    var answer = MessageBox.Show(
                        "A newer regex list is available (v" + remoteRules.Version
                        + "). Replace your current rules with the updated defaults?",
                        "Sanity — regex update",
                        MessageBoxButtons.YesNo,
                        MessageBoxIcon.Question);

                    if (answer == DialogResult.Yes)
                    {
                        config.Rules = remoteRules.Rules;
                        config.RulesVersion = remoteRules.Version;
                    }
                }

                var localDate = File.GetLastWriteTimeUtc(Assembly.GetExecutingAssembly().Location);
                if (remoteReleaseDate.HasValue && remoteReleaseDate.Value > localDate.AddMinutes(1))
                {
                    var answer = MessageBox.Show(
                        "A newer Sanity build is available on GitHub"
                        + " (release file dated "
                        + remoteReleaseDate.Value.ToLocalTime().ToString("d", CultureInfo.CurrentCulture)
                        + "). Open the download page?",
                        "Sanity — app update",
                        MessageBoxButtons.YesNo,
                        MessageBoxIcon.Question);

                    if (answer == DialogResult.Yes)
                        AppLinks.Open(AppLinks.ReleaseAsset);
                }
            }
            finally
            {
                config.LastUpdateCheck = DateTime.UtcNow;
                config.Save();
            }
        }

        private static DateTime? FetchReleaseFileDate()
        {
            EnsureTls();
            using (var client = new WebClient())
            {
                client.Headers[HttpRequestHeader.UserAgent] = "Sanity";
                client.Headers[HttpRequestHeader.Accept] = "application/vnd.github+json";
                var json = client.DownloadString(AppLinks.ReleaseCommitsApi);
                var commits = new JavaScriptSerializer().DeserializeObject(json) as object[];
                if (commits == null || commits.Length == 0)
                    return null;

                var first = commits[0] as System.Collections.Generic.Dictionary<string, object>;
                if (first == null || !first.ContainsKey("commit"))
                    return null;

                var commit = first["commit"] as System.Collections.Generic.Dictionary<string, object>;
                if (commit == null || !commit.ContainsKey("committer"))
                    return null;

                var committer = commit["committer"] as System.Collections.Generic.Dictionary<string, object>;
                if (committer == null || !committer.ContainsKey("date"))
                    return null;

                var dateText = committer["date"] as string;
                if (string.IsNullOrEmpty(dateText))
                    return null;

                DateTime parsed;
                if (!DateTime.TryParse(
                    dateText,
                    CultureInfo.InvariantCulture,
                    DateTimeStyles.RoundtripKind,
                    out parsed))
                    return null;

                return parsed.ToUniversalTime();
            }
        }

        private static void EnsureTls()
        {
            try
            {
                ServicePointManager.SecurityProtocol |= (SecurityProtocolType)3072;
            }
            catch
            {
            }
        }
    }
}

File: mac/Sanity/Sources/UpdateChecker.swift
import AppKit
import Foundation

enum UpdateChecker {
    private static let checkInterval: TimeInterval = 7 * 24 * 60 * 60
    private static var running = false
    private static let lock = NSLock()

    static func runAsync(config: AppConfig, apply: @escaping (AppConfig) -> Void) {
        guard beginIfDue(config) else { return }

        DispatchQueue.global(qos: .utility).async {
            let remoteRules = (try? DefaultRules.fetchRemote())
            let remoteReleaseDate = (try? fetchReleaseFileDate())

            DispatchQueue.main.async {
                var updated = config
                prompt(config: &updated, remoteRules: remoteRules, remoteReleaseDate: remoteReleaseDate)
                updated.lastUpdateCheck = AppConfig.encodeSleepDate(Date())
                updated.save()
                endRunning()
                apply(updated)
            }
        }
    }

    /// Blocks while fetching and prompting — for short-lived --open processes.
    static func runSync(config: AppConfig) {
        guard beginIfDue(config) else { return }
        defer { endRunning() }

        let remoteRules = (try? DefaultRules.fetchRemote())
        let remoteReleaseDate = (try? fetchReleaseFileDate())

        var updated = config
        prompt(config: &updated, remoteRules: remoteRules, remoteReleaseDate: remoteReleaseDate)
        updated.lastUpdateCheck = AppConfig.encodeSleepDate(Date())
        updated.save()
    }

    private static func beginIfDue(_ config: AppConfig) -> Bool {
        guard config.updatesEnabled, isDue(config) else { return false }
        lock.lock()
        defer { lock.unlock() }
        if running { return false }
        running = true
        return true
    }

    private static func endRunning() {
        lock.lock()
        running = false
        lock.unlock()
    }

    private static func isDue(_ config: AppConfig) -> Bool {
        guard let raw = config.lastUpdateCheck, let last = parseDate(raw) else {
            return true
        }
        return Date().timeIntervalSince(last) >= checkInterval
    }

    private static func parseDate(_ value: String) -> Date? {
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = iso.date(from: value) { return date }
        iso.formatOptions = [.withInternetDateTime]
        return iso.date(from: value)
    }

    private static func prompt(
        config: inout AppConfig,
        remoteRules: RegexRulesCatalog?,
        remoteReleaseDate: Date?
    ) {
        if let remoteRules, remoteRules.version > config.rulesVersion {
            let alert = NSAlert()
            alert.messageText = "Regex update"
            alert.informativeText =
                "A newer regex list is available (v\(remoteRules.version)). Replace your current rules with the updated defaults?"
            alert.addButton(withTitle: "Update")
            alert.addButton(withTitle: "Not now")
            if alert.runModal() == .alertFirstButtonReturn {
                config.rules = remoteRules.rules
                config.rulesVersion = remoteRules.version
            }
        }

        let localDate = executableDate()
        if let remoteReleaseDate, let localDate, remoteReleaseDate > localDate.addingTimeInterval(60) {
            let formatter = DateFormatter()
            formatter.dateStyle = .medium
            let alert = NSAlert()
            alert.messageText = "App update"
            alert.informativeText =
                "A newer Sanity build is available on GitHub (release file dated \(formatter.string(from: remoteReleaseDate))). Open the download page?"
            alert.addButton(withTitle: "Open")
            alert.addButton(withTitle: "Not now")
            if alert.runModal() == .alertFirstButtonReturn {
                AppLinks.open(AppLinks.releaseAsset)
            }
        }
    }

    private static func executableDate() -> Date? {
        guard let path = Bundle.main.executableURL?.path else { return nil }
        let attrs = try? FileManager.default.attributesOfItem(atPath: path)
        return attrs?[.modificationDate] as? Date
    }

    private static func fetchReleaseFileDate() throws -> Date {
        let url = URL(string: AppLinks.releaseCommitsApi)!
        var request = URLRequest(url: url, timeoutInterval: 20)
        request.setValue("Sanity", forHTTPHeaderField: "User-Agent")
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")

        let semaphore = DispatchSemaphore(value: 0)
        var resultData: Data?
        var resultError: Error?
        URLSession.shared.dataTask(with: request) { data, _, error in
            resultData = data
            resultError = error
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let resultError { throw resultError }
        guard let resultData else {
            throw NSError(domain: "Sanity", code: 4, userInfo: [NSLocalizedDescriptionKey: "Empty commits response."])
        }

        struct CommitItem: Decodable {
            struct Commit: Decodable {
                struct Committer: Decodable { let date: Date }
                let committer: Committer
            }
            let commit: Commit
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let items = try decoder.decode([CommitItem].self, from: resultData)
        guard let date = items.first?.commit.committer.date else {
            throw NSError(domain: "Sanity", code: 5, userInfo: [NSLocalizedDescriptionKey: "No commit date."])
        }
        return date
    }
}
